### PR TITLE
[FDB] Fix fbdorch to properly handle syncd FDB FLUSH Notif

### PR DESCRIPTION
**Pull request by vivekreddynv**
**What I did**

Fixes https://github.com/Azure/sonic-buildimage/issues/10605

**Why I did it**

1) Vlan delete couldn't be handled by OA when there is fdb learnt on the member and when the member is deleted
2) This inability of handling APPL_DB notif is affecting warm-restart
3) FDB Entry from State DB is not removed.
4) OA doesn't have the logic to handle consolidate flush notif coming from syncd
5) FdbOrch doesn't have logic to clear internal cache and decrement corresponding fdb counters during a flush notification

**How I verified it**

Checked this config and verified the remove vlan passed though
```
1. config vlan add 40
2. config vlan member add 40 Ethernet0
On the other side. 
3. sudo ip link add link enp4s0f0 name enp4s0f0.40 type vlan id 40
4. sudo ip link set enp4s0f0.40 up
5. Make sure a fdb entry is learnt. 

root@sonic:/home/admin# fdbshow
  No.    Vlan  MacAddress         Port       Type
-----  ------  -----------------  ---------  -------
    1      40  7C:FE:90:12:22:EC  Ethernet0  Dynamic

dump state fdb all --db STATE_DB
{
    "Vlan40:7c:fe:90:12:22:ec": {
        "STATE_DB": {
            "keys": [
                {
                    "FDB_TABLE|Vlan40:7c:fe:90:12:22:ec": {
                        "port": "Ethernet0",
                        "type": "dynamic"
                    }
                }
            ],
            "tables_not_found": []
        }
    }
}

7. config vlan member del 40 Ethernet0
8. config vlan del 40
9. dump state fdb all --db STATE_DB
{}
```
**Unit Tests:**
```
vkarri@4897dcba242c:/sonic/src/sonic-swss/tests/mock_tests$ ./tests --gtest_filter=FdbOrchTest*
Running main() from /usr/src/gtest/src/gtest_main.cc
Note: Google Test filter = FdbOrchTest*
[==========] Running 6 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 6 tests from FdbOrchTest
[ RUN      ] FdbOrchTest.ConsolidatedFlushVlanandPort
[       OK ] FdbOrchTest.ConsolidatedFlushVlanandPort (14 ms)
[ RUN      ] FdbOrchTest.ConsolidatedFlushAll
[       OK ] FdbOrchTest.ConsolidatedFlushAll (15 ms)
[ RUN      ] FdbOrchTest.ConsolidatedFlushVlan
[       OK ] FdbOrchTest.ConsolidatedFlushVlan (14 ms)
[ RUN      ] FdbOrchTest.ConsolidatedFlushPort
[       OK ] FdbOrchTest.ConsolidatedFlushPort (15 ms)
[ RUN      ] FdbOrchTest.ConsolidatedFlushVlanandPortBridgeportDeleted
[       OK ] FdbOrchTest.ConsolidatedFlushVlanandPortBridgeportDeleted (14 ms)
[ RUN      ] FdbOrchTest.NonConsolidatedFlushVlanandPort
[       OK ] FdbOrchTest.NonConsolidatedFlushVlanandPort (15 ms)
[----------] 6 tests from FdbOrchTest (87 ms total)

[----------] Global test environment tear-down
[==========] 6 tests from 1 test case ran. (87 ms total)
[  PASSED  ] 6 tests.

```

**Details if related**

### CODE REVIEW

**Review by dgsudharsan — APPROVED** [review]

**Review by dgsudharsan — APPROVED** [review]

File: orchagent/fdborch.cpp [hunk 1]
```diff
@@ -174,41 +173,160 @@ bool FdbOrch::storeFdbEntryState(const FdbUpdate& update)
...
+                                    const MacAddress& mac)
+{
+    MacAddress flush_mac(CONSOLIDATED_FLUSH_MAC);
+
+    if (bridge_port_id == SAI_NULL_OBJECT_ID && bv_id == SAI_NULL_OBJECT_ID)
+    {
+        for (auto itr = m_entries.begin(); itr != m_entries.end();)
+        {
+            auto curr = itr++;
+            if (curr->second.type.find(fdb_type) != std::string::npos)
```
Comment (prsunny) [inline orchagent/fdborch.cpp:224]: I think this check is redundant. afaik, the fdb_type is always populated.

File: orchagent/fdborch.cpp [hunk 2]
```diff
@@ -174,41 +173,160 @@ bool FdbOrch::storeFdbEntryState(const FdbUpdate& update)
...
+    storeFdbEntryState(update);
+    notify(SUBJECT_TYPE_FDB_CHANGE, &update);
+
+    SWSS_LOG_INFO("FdbEntry removed from internal cache, MAC: %s , port: %s, BVID: 0x%" PRIx64,
+                   mac.to_string().c_str(), port_alias.c_str(), bv_id);
+}
+
+/*
+Handles the SAI_FDB_EVENT_FLUSHED notification recieved from syncd
+Ref: https://github.com/opencomputeproject/SAI/blob/master/inc/saifdb.h#L223
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply orchagent/fdborch.cpp]: Handled

File: orchagent/fdborch.cpp [hunk 3]
```diff
@@ -174,41 +173,160 @@ bool FdbOrch::storeFdbEntryState(const FdbUpdate& update)
...
                         bridge_port_id);
-
+            return;
         }
+    }
+
+    if (entry->bv_id &&
+        !m_portsOrch->getPort(entry->bv_id, vlan))
+    {
+        SWSS_LOG_ERROR("FdbOrch notification type %d: Failed to locate vlan port from bv_id 0x%" PRIx64, type, entry->bv_id);
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply orchagent/fdborch.cpp:330]: Handled

File: orchagent/fdborch.cpp (hunk 1 above)
Comment (vivekreddynv) [reply orchagent/fdborch.cpp:225]: These entries might be static as well, so i think it's required to check the fdb_type before clearing them

File: orchagent/fdborch.cpp [hunk 4]
```diff
@@ -493,86 +594,21 @@ void FdbOrch::update(sai_fdb_event_t        type,
     }
     case SAI_FDB_EVENT_FLUSHED:
 
-        SWSS_LOG_INFO("FDB Flush event received: [ %s , 0x%" PRIx64 " ], \
-                       bridge port ID: 0x%" PRIx64 ".",
+        SWSS_LOG_NOTICE("FDB Flush event received: [ %s , 0x%" PRIx64 " ], \
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply orchagent/fdborch.cpp]: Handled

File: orchagent/fdborch.cpp [hunk 5]
```diff
@@ -174,41 +173,160 @@ bool FdbOrch::storeFdbEntryState(const FdbUpdate& update)
...
+/*
+Handles the SAI_FDB_EVENT_FLUSHED notification recieved from syncd
+Ref: https://github.com/opencomputeproject/SAI/blob/master/inc/saifdb.h#L223
+*/
+void FdbOrch::handleSyncdFlushNotif(const sai_object_id_t& bv_id,
+                                    const sai_object_id_t& bridge_port_id,
+                                    const string& fdb_type,
+                                    const MacAddress& mac)
+{
+    MacAddress flush_mac(CONSOLIDATED_FLUSH_MAC);
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply orchagent/fdborch.cpp]: Handled

File: orchagent/fdborch.cpp (hunk 1 above)
Comment (prsunny) [reply orchagent/fdborch.cpp:225]: so you are not checking for static vs dynamic here. my point is, one or other is always true and never be npos, right?

Comment (prsunny) [reply orchagent/fdborch.cpp:225]: May be i'm missing something. what do we miss with the following?

```
            if (curr->second.type.find(fdb_type) != std::string::npos)
            {
                if (curr->first.mac == mac || mac == flush_mac)
                {
                    clearFdbEntry(curr->first.mac, curr->first.bv_id, fdb_type, curr->first.port_name);
                }
            }
```

to 
```

                if (curr->first.mac == mac || mac == flush_mac)
                {
                    clearFdbEntry(curr->first.mac, curr->first.bv_id, fdb_type, curr->first.port_name);
                }
```

Comment (vivekreddynv) [reply orchagent/fdborch.cpp:225]: We're basically checking if the entry type stored in m_entries is equivalent to what was received by flush. 

i.e. only proceed forward in these (static == static || dynamic == dynamic || dynamic == dynamic_local) scenarios

Comment (prsunny) [reply orchagent/fdborch.cpp:225]: Got it. In the flush, can user specify fdb_type? If not, is it correct that flush notification would be for all types?

**Review by dgsudharsan — APPROVED** [review]

File: orchagent/fdborch.cpp (hunk 1 above)
Comment (vivekreddynv) [reply orchagent/fdborch.cpp:225]: [Link](https://github.com/opencomputeproject/SAI/blob/master/inc/saifdb.h#L222) User can specify the flush type. Looks like there can also be a SAI_FDB_FLUSH_ENTRY_TYPE_ALL along with SAI_FDB_FLUSH_ENTRY_TYPE_STATIC/SAI_FDB_FLUSH_ENTRY_TYPE_DYNAMIC

Comment (vivekreddynv) [reply orchagent/fdborch.cpp:225]: I'll update to handle the SAI_FDB_FLUSH_ENTRY_TYPE_ALL case as well.

Comment (vivekreddynv) [reply orchagent/fdborch.cpp:225]: Sry, the flush notif will have either STATIC or DYNAMIC, so "all" case need not be handled.

Comment (vivekreddynv) [reply orchagent/fdborch.cpp]: EDIT: Updated, remove reading the SAI_FDB_ENTRY_ATTR_TYPE attribute and cleared only non-static entries

**Review by prsunny — APPROVED** [review]

**Pull request merged**